django/contrib/contenttypes/fields.py

```diff
@@ -564,19 +568,28 @@ def get_prefetch_queryset(self, instances, queryset=None):
 
             queryset._add_hints(instance=instances[0])
             queryset = queryset.using(queryset._db or self._db)
-
-            query = {
-                '%s__pk' % self.content_type_field_name: self.content_type.id,
-                '%s__in' % self.object_id_field_name: {obj.pk for obj in instances}
-            }
-
+            # Group instances by content types.
+            content_type_queries = (
+                models.Q(**{
+                    '%s__pk' % self.content_type_field_name: content_type_id,
+                    '%s__in' % self.object_id_field_name: {obj.pk for obj in objs}
+                })
+                for content_type_id, objs in itertools.groupby(
+                    sorted(instances, key=lambda obj: self.get_content_type(obj).pk),
+                    lambda obj: self.get_content_type(obj).pk,
```

Comment:
This seems rather inefficient... Doesn't `self.get_content_type(obj)` execute a query?

Reply:
@pope1ni no, it's heavily cached as it's using `ContentType.objects.get_for_model`.